src/adaptor/ini.rs

```diff
@@ -0,0 +1,502 @@
+
+        // parse the basic INI structure
+        let output = match sections(&buffer) {
+            IResult::Done(_, o) => o,
+            _ => return Err("unable to parse INI data"),
+        };
+
+        let mut combined = HashMap::new();
+
+        // Here we convert the INI into our configuration AST,
+        // performing section and key de-duplication as necessary
+        for (name, pairs) in output {
+            // fetch existing entry or create a new one, deduplicating sections
+            let mut entry = combined.entry(name.into()).or_insert_with(|| HashMap::new());
+            // later, we will need schema data in order to encode type information into the AST
+            // for now, just assume everything is a string
+            let converted = pairs.iter().map(|&(key,value)| (key.to_string(), Config::Text(value.to_string())));
+            insert_all(entry, converted);
+        }
+
+        // wrap it all up in an object
+        let objects = combined.into_iter().map(|(key, value)| (key, Config::Object(value))).collect();
+        Ok(Config::Object(objects))
+    }
+
+    /// Serialize the `Config` AST into INI format
+    fn serialize<W>(&self, config: Config, mut writer: W) -> Result<(), AdaptorError> 
+        where W: Write 
+    {
+        let ini_model = try!(convert_model(config));
+
+        for (header, props) in ini_model {
+            writeln!(writer, "[{}]", header);
+            for (key, value) in props {
+                writeln!(writer, "{} = {}", key, value);
+            }
+            writeln!(writer, "");
+        }
+
+        Ok(())
+    }
+}
+
+#[test]
+fn deserialize_ini_section() {
+    let adaptor = IniAdaptor::new();
+    let mut ini = b"[section]
+key = value";
+
+    let config = adaptor.deserialize(&ini[..]).unwrap();
+    let mut pairs = HashMap::new();
+    pairs.insert("key".to_string(), Config::Text("value".to_string()));
+    let mut sections = HashMap::new();
+    sections.insert("section".to_string(), Config::Object(pairs));
+    assert_eq!(config, Config::Object(sections));
+
+}
+
+#[test]
+fn deserialize_ini_duplicate_keys() {
+    let adaptor = IniAdaptor::new();
+    let mut ini = b"[section]
+key = value1
+key = value2";
+
+    let config = adaptor.deserialize(&ini[..]).unwrap();
+    let mut pairs = HashMap::new();
+    pairs.insert("key".to_string(), Config::Array(
+        vec![
+            Config::Text("value1".to_string()),
+            Config::Text("value2".to_string()),
+        ]
+    ));
+    let mut sections = HashMap::new();
+    sections.insert("section".to_string(), Config::Object(pairs));
+    assert_eq!(config, Config::Object(sections));
+
+}
+
+#[test]
+fn serialize_ini_section() {
+    let adaptor = IniAdaptor::new();
+
+    let mut pairs = HashMap::new();
+    pairs.insert("key".to_string(), Config::Text("value".to_string()));
+    let mut section = HashMap::new();
+    section.insert("section".to_string(), Config::Object(pairs));
```

Comment:
I'll add tests after I add the ability to parse key-value pairs without a section in the next PR.